File: go/orm/internal/testsupport/sample_database.go
package testsupport

// This file may import orm, sample, and sqlite: orm's own tests live in an
// external test package (package orm_test), so there is no import cycle.

import (
	"context"
	"testing"
	"time"

	"github.com/kosmassp/SimpleOrm/go/orm"
	"github.com/kosmassp/SimpleOrm/go/orm/sample"
	"github.com/kosmassp/SimpleOrm/go/orm/sqlite"
)

// SeedTime is the fixed instant test fixtures use for created_at (mirrors
// dotnet's TestDb.SeedTime / PHP's SampleDatabase).
var SeedTime = time.Date(2026, 8, 28, 9, 30, 0, 0, time.UTC)

// OpenSample opens a temp-file database with every sample table and the
// UserTransactionTotal view created from metadata (ADR-0011, ADR-0008 add.3) —
// no hand-written DDL and no dependency on the migrations runner, which lands
// in a parallel phase of this port. Closed on test cleanup.
func OpenSample(t testing.TB) *orm.Db {
	t.Helper()
	ctx := context.Background()

	db, err := orm.Open(ctx, TempDatabase(t), orm.Options{Dialect: sqlite.New()})
	if err != nil {
		t.Fatalf("open sample database: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })

	for _, create := range []func(context.Context, *orm.Db) error{
		orm.CreateTable[sample.User],
		orm.CreateTable[sample.Role],
		orm.CreateTable[sample.UserRole],
		orm.CreateTable[sample.UserProfile],
		orm.CreateTable[sample.Transaction],
		orm.CreateTable[sample.TransactionDetail],
	} {
		if err := create(ctx, db); err != nil {
			t.Fatalf("create sample table: %v", err)
		}
	}
	if err := orm.CreateView[sample.UserTransactionTotal](ctx, db); err != nil {
		t.Fatalf("create sample view: %v", err)
	}

	return db
}

// InsertUser inserts a User with SeedTime as its created_at (mirrors dotnet's
// TestDb.InsertUserAsync / PHP's SampleDatabase::insertUser).
func InsertUser(t testing.TB, db *orm.Db, name, email string) *sample.User {
	t.Helper()
	user := &sample.User{Name: name, Email: email}
	user.CreatedAtUtc = SeedTime
	if err := orm.Insert(context.Background(), db, user); err != nil {
		t.Fatalf("insert user: %v", err)
	}
	return user
}

File: php/tests/Support/SampleDatabase.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Support;

use DateTimeImmutable;
use SimpleOrm\Dialect\SqliteDialect;
use SimpleOrm\Session\Db;
use SimpleOrm\Session\DbOptions;
use SimpleOrm\Tests\Sample\Models\Role;
use SimpleOrm\Tests\Sample\Models\Transaction;
use SimpleOrm\Tests\Sample\Models\TransactionDetail;
use SimpleOrm\Tests\Sample\Models\User;
use SimpleOrm\Tests\Sample\Models\UserProfile;
use SimpleOrm\Tests\Sample\Models\UserRole;
use SimpleOrm\Tests\Sample\Models\UserTransactionTotal;
use SimpleOrm\Types\Decimal;

/**
 * Session setup for Session/Query/Mapping integration tests (mirrors dotnet's
 * `TestDb`): opens a {@see Db} on a temp file and creates the whole sample
 * schema — tables and the view — from entity metadata (ADR-0011 + ADR-0008
 * add.3). No hand-written DDL, and no dependency on the migrations tree (a
 * separate, in-flux area of this port).
 */
final class SampleDatabase
{
    /** The fixed instant every sample fixture inserts with, matching conformance/fixtures/seed.json. */
    public static function seedTime(): DateTimeImmutable
    {
        return new DateTimeImmutable('2026-08-28T09:30:00.0000000Z');
    }

    public static function options(): DbOptions
    {
        return new DbOptions(new SqliteDialect());
    }

    public static function open(TempDatabase $fixture): Db
    {
        $db = Db::open($fixture->connectionString(), self::options());

        $db->createTable(User::class);
        $db->createTable(Role::class);
        $db->createTable(UserRole::class);
        $db->createTable(UserProfile::class);
        $db->createTable(Transaction::class);
        $db->createTable(TransactionDetail::class);
        $db->createView(UserTransactionTotal::class);

        return $db;
    }

    public static function insertUser(Db $db, string $name, string $email): User
    {
        $user = new User();
        $user->name = $name;
        $user->email = $email;
        $user->createdAtUtc = self::seedTime();
        $db->insert($user);

        return $user;
    }

    /** A transaction owned by `$user` (Level 2 eager-loading fixtures: EagerJoinTest). */
    public static function insertTransaction(Db $db, User $user, string $amount): Transaction
    {
        $transaction = new Transaction();
        $transaction->userId = $user->id;
        $transaction->amount = Decimal::of($amount);
        $transaction->createdAtUtc = self::seedTime();
        $db->insert($transaction);

        return $transaction;
    }

    /** Level 2 eager-loading fixtures (EagerJoinTest): a role, not yet granted to anyone. */
    public static function insertRole(Db $db, string $name): Role
    {
        $role = new Role();
        $role->name = $name;
        $role->createdAtUtc = self::seedTime();
        $db->insert($role);

        return $role;
    }

    /** Level 2 eager-loading fixtures (EagerJoinTest): grants `$role` to `$user` via the `UserRole` link. */
    public static function insertUserRole(Db $db, User $user, Role $role): UserRole
    {
        $userRole = new UserRole();
        $userRole->userId = $user->id;
        $userRole->roleId = $role->id;
        $userRole->createdAtUtc = self::seedTime();
        $db->insert($userRole);

        return $userRole;
    }

    /** Level 2 eager-loading fixtures (EagerJoinTest): the 1:1 profile side, FK on `UserProfile`. */
    public static function insertProfile(Db $db, User $user, ?string $bio = null): UserProfile
    {
        $profile = new UserProfile();
        $profile->userId = $user->id;
        $profile->bio = $bio;
        $profile->createdAtUtc = self::seedTime();
        $db->insert($profile);

        return $profile;
    }
}
